include/cuda/std/detail/libcxx/include/atomic

```diff
@@ -1243,269 +1243,282 @@ _LIBCUDACXX_INLINE_VISIBILITY void __cxx_atomic_wait(__cxx_atomic_impl<_Tp, _Sco
         if (memory_order_acq_rel == __m)
-            return __cxx_atomic_compare_exchange_strong(&__a_, &__e, __d, __m, memory_order_acquire);
+            return __cxx_atomic_compare_exchange_strong(&this->__a_, &__e, __d, __m, memory_order_acquire);
         else if (memory_order_release == __m)
-            return __cxx_atomic_compare_exchange_strong(&__a_, &__e, __d, __m, memory_order_relaxed);
+            return __cxx_atomic_compare_exchange_strong(&this->__a_, &__e, __d, __m, memory_order_relaxed);
         else
-            return __cxx_atomic_compare_exchange_strong(&__a_, &__e, __d, __m, __m);
+            return __cxx_atomic_compare_exchange_strong(&this->__a_, &__e, __d, __m, __m);
     }
 
     _LIBCUDACXX_INLINE_VISIBILITY void wait(_Tp __v, memory_order __m = memory_order_seq_cst) const volatile _NOEXCEPT
-        {__cxx_atomic_wait(&__a_, __v, __m);}
+        {__cxx_atomic_wait(&this->__a_, __v, __m);}
     _LIBCUDACXX_INLINE_VISIBILITY void wait(_Tp __v, memory_order __m = memory_order_seq_cst) const _NOEXCEPT
-        {__cxx_atomic_wait(&__a_, __v, __m);}
+        {__cxx_atomic_wait(&this->__a_, __v, __m);}
     _LIBCUDACXX_INLINE_VISIBILITY void notify_one() volatile _NOEXCEPT
-        {__cxx_atomic_notify_one(&__a_);}
+        {__cxx_atomic_notify_one(&this->__a_);}
     _LIBCUDACXX_INLINE_VISIBILITY void notify_one() _NOEXCEPT
-        {__cxx_atomic_notify_one(&__a_);}
+        {__cxx_atomic_notify_one(&this->__a_);}
     _LIBCUDACXX_INLINE_VISIBILITY void notify_all() volatile _NOEXCEPT
-        {__cxx_atomic_notify_all(&__a_);}
+        {__cxx_atomic_notify_all(&this->__a_);}
     _LIBCUDACXX_INLINE_VISIBILITY void notify_all() _NOEXCEPT
-        {__cxx_atomic_notify_all(&__a_);}
+        {__cxx_atomic_notify_all(&this->__a_);}
 };
 
-template <class _Tp, int _Sco = 0, bool = (is_integral<_Tp>::value || is_floating_point<_Tp>::value) && !is_same<_Tp, bool>::value>
-struct __atomic_base_ref {
-    mutable __cxx_atomic_ref_impl<_Tp, _Sco> __a_;
+template <class _Tp, typename _Storage>
+struct __atomic_base_core<_Tp, true, _Storage> : public __atomic_base_storage<_Tp, _Storage>{
+    __atomic_base_core() = default;
+    __atomic_base_core(const __atomic_base_core&) = default;
+    __atomic_base_core(__atomic_base_core&&) = default;
 
-    __atomic_base_ref() = default;
-    __atomic_base_ref(const __atomic_base_ref&) = default;
-    __atomic_base_ref(__atomic_base_ref&&) = default;
-
-    __atomic_base_ref& operator=(const __atomic_base_ref&) = delete;
-    __atomic_base_ref& operator=(__atomic_base_ref&&) = delete;
+    __atomic_base_core& operator=(const __atomic_base_core&) = default;
+    __atomic_base_core& operator=(__atomic_base_core&&) = default;
 
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR
-    __atomic_base_ref(_Tp& __a) _NOEXCEPT : __a_(__a) {}
+    __atomic_base_core(_Storage&& __a) _NOEXCEPT : __atomic_base_storage<_Tp, _Storage>(forward<_Storage>(__a)) {}
```

Comment:
technically we are required to qualify `forward` as it is non ugly function

Reply:
Good catch.